---
title: "Bats_counttable"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#import libraries to be able to build table count for each gene
library("tximport") #for tximport
library(DESeq2) #for the differential expression analysis
library(biomaRt) #to automatically request ensembl identifiers
library(dplyr) #for the manipulation of data 
library(tidyr)
library(data.table)
library(ggplot2) #for graphs

#set working directory
setwd("/ifb/data/mydatalocal/data_from_ftp")
```

# Differential Expression analysis 

Import the files, we need a variable that contains the name of the files quant.sf
```{r}
dir <- "/ifb/data/mydatalocal/data_from_ftp/outputs_salmon" #the directory where we look for the files
all_files <- list.files(dir)

#get the quant.sf files for all samples
#initialisation
files_Lib <- c()
names_file <- c()
i <- 1

#loop to get all the files in the same place
for (name in all_files) {
    if (grepl("salmon_Lib",as.character(name))==TRUE) {
      files_Lib[i] <-  file.path(dir,name,"quant.sf", fsep="/")
      names_file[i] <- c(name)
      i <- i+1
      }
  }

names(files_Lib) <- names_file #give them names
```

Now we build reading table from the trinity gene trans map, to have the correspondance between gene and transcript
```{r}
Tab_trini <- read.table("/ifb/data/mydatalocal/data_from_ftp/trinity_data/Trinity_RF.fasta.gene_trans_map") #get the fasta.gene_trans_map
names(Tab_trini) <- c("GENEID","TXNAME") #to have the right column names
Tab_trini <- Tab_trini[,c(2,1)] #to have the right order: TXNAME as first column, GENEID as second
```

Now let's build a table with the count from each gene from the outputs of salmon 
```{r}
txi <- tximport(files_Lib, type = "salmon", tx2gene = Tab_trini)
names(txi) #outputs, we obtained 3 tables: abundance, counts, and length
head(txi$counts) #what's interesting for DESeq is the counts
```

Now let's use DESeq
```{r}
#build table of correspondence which sample is + or - IFN
samp.name <- colnames(txi$counts)
samp.type <- factor(rep(c("CTL", "IFN"),each= 3)) #3 first are control, 3 last are +IFN
samples <- data.frame(run=samp.name,condition=samp.type)

#now build DESeq object (dds)
ddsTxi <- DESeqDataSetFromTximport(txi, colData = samples, design = ~ condition)

#Run DEseq
dds <- DESeq(ddsTxi)
res <- results(dds)
res
#save the table
#write.table(res,"/ifb/data/mydatalocal/data_from_ftp/DESeq/DESeqRes.tab")

#count number of significative p-values
table(res$padj <0.05) #number of tests that are under the threshold of 0.05 #1457 
```

Now we do control quality, and verify that there is no abnormality with the replicates (samples from the same conditions have to be closer than different conditions)
```{r}
#MA-plot 
plotMA(res, ylim=c(-2,2))

#PCA
vsd <- vst(dds, blind=FALSE) #DESeq transform (variance stabilizing transformation)
plotPCA(vsd, intgroup=c("condition","run"))
#The Lib3 (CTL) is an outlier, and is not grouped with the other samples from the same condition. It can be explained because the RNA library was built a different day from the 2 others. -> We redo the DESeq analysis only on Lib1,2,4,5 and 6 (cf Bats_DEseq_without3.rmd)

#To have an idea of what the results can look like for a gene, let's plot the counts for one gene (we choose the one with the lowest adjusted p value)
plotCounts(dds, gene=which.min(res$padj), intgroup="condition")
```

Now we want to cross the results of Blast with the outputs from DEseq, in order to identify the names of all the genes (based on their corresponding human CDS)
```{r}
#Import Blast table
Blast <- read.table("/ifb/data/mydatalocal/data_from_ftp/blast_data/DataBats.blast")

#now we make the correspondence between ENSEMBL identifiers and external human gene names, using biomaRt
#match gene/transcript names
ensembl <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")
tx2geneHomo = biomaRt::getBM(attributes=c('ensembl_transcript_id', 
                                      'ensembl_gene_id', 'external_gene_name'),mart=ensembl)
#add gene name to the blast outputs
blastHomo<-data.table(Blast)
blastHomo<-blastHomo[, c("trid", "version") := tstrsplit(V2, ".", fixed=TRUE)][]
blastHomo<-blastHomo[, c("gene", "alttr") := tstrsplit(V1, "_i", fixed=TRUE)][]

blastHomo1 <- blastHomo%>% group_by(gene)%>% slice_max(order_by = V12, n = 1, with_ties = TRUE)

blastHomoName<-merge(tx2geneHomo,blastHomo1,by.x="ensembl_transcript_id",by.y="trid") #to merge the 2 dataframe, one with the Blast output and ensembl names (blastHomo1), and one with the ensemble and external gene names (tx2gneHomo)

blastHomoNameUniq<-unique(blastHomoName[,c("ensembl_gene_id","external_gene_name","gene")]) #unique, to remove the duplicates

#associate our DE results with gene names from human
resUsName=merge(res,blastHomoNameUniq,by.x=0,by.y="gene")
#view(ResUsName)
#save this table
#write.csv(resUsName,file="/ifb/data/mydatalocal/data_from_ftp/DESeq/Tab_foldchange_genes.csv")
```

# Analyze the data
Number of genes up and down regulated
```{r}
table(resUsName$padj <0.05) #Number of genes identified as significatively differently expressed (with inteferon vs without) (578)
table(resUsName$padj <0.05 & resUsName$log2FoldChange>0) #Number of genes significatively UPregulated (421)
table(resUsName$padj <0.05 & resUsName$log2FoldChange<0) #Number of genes significatively DOWNregulated (157)

#filter the downregulated and upregulated genes
UpReg <- filter(resUsName, padj<0.05 & resUsName$log2FoldChange>0)
DownReg <- filter(resUsName, padj<0.05 & resUsName$log2FoldChange<0)
AllReg <- filter(resUsName, padj<0.05)
```
Export files for Gene Ontology analysis 
```{r}
#export files with gene names for Gene Ontology analysis (GOrilla)
write.csv(as.data.frame(resUsName$external_gene_name),file="/ifb/data/mydatalocal/data_from_ftp/DESeq/name_all.csv",quote=FALSE,row.names = FALSE) #background set, all genes found in our data
write.csv(as.data.frame(UpReg$external_gene_name),file="/ifb/data/mydatalocal/data_from_ftp/DESeq/name_up.csv",quote=FALSE,row.names = FALSE) #all upregulated genes
write.csv(as.data.frame(DownReg$external_gene_name),file="/ifb/data/mydatalocal/data_from_ftp/DESeq/name_down.csv",quote=FALSE,row.names = FALSE) #all downregulated genes
write.csv(as.data.frame(AllReg$external_gene_name),file="/ifb/data/mydatalocal/data_from_ftp/DESeq/name_all_reg.csv",quote=FALSE,row.names = FALSE) #all regulated genes (up and down)
```
Compare our results with existing results
```{r}
#Compare with publication Holzer  (+6h IFN)
#import their data
resBatHolzer<-read.csv("https://www.rna.uni-jena.de/supplements/mda/data/deseq2/RNA/Mock_6h_RNA_vs_IFN_6h_RNA/deseq2_Mock_6h_RNA_IFN_6h_RNA_full.csv")
colnames(resBatHolzer)[2] <- c("external_gene_name") #change column name to be the same as ours

table(resBatHolzer$padj <0.05) #Number of genes identified as significatively differently expressed in their study (with interferon vs without) (1702)
table(resBatHolzer$padj <0.05 & resBatHolzer$log2FoldChange>0) #Number of genes significatively UPregulated (1097)
table(resBatHolzer$padj <0.05 & resBatHolzer$log2FoldChange<0) #Number of genes significatively DOWNregulated (605)

#Genes in common
CommonGenesUsHolze <- merge(resUsName,resBatHolzer,by="external_gene_name") #11364 genes in common

#Chi2 test

cont_tb<-table(us=CommonGenesUsHolze$padj.x<0.05,them=CommonGenesUsHolze$padj.y < 0.05) #Contingency table
chisq.test(cont_tb) #p< 2.2e-16

#add a column for our test to mark if gene is up-regulated, down-regulated or not regulated
CommonGenesUsHolze$ourTest="nonreg" #by default
CommonGenesUsHolze$ourTest[CommonGenesUsHolze$padj.x<0.05&CommonGenesUsHolze$log2FoldChange.x>0]="Up"
CommonGenesUsHolze$ourTest[CommonGenesUsHolze$padj.x<0.05&CommonGenesUsHolze$log2FoldChange.x<0]="Down"
#idem for Holzer's test
CommonGenesUsHolze$theirTest="nonreg"
CommonGenesUsHolze$theirTest[CommonGenesUsHolze$padj.y<0.05&CommonGenesUsHolze$log2FoldChange.y>0]="Up"
CommonGenesUsHolze$theirTest[CommonGenesUsHolze$padj.y<0.05&CommonGenesUsHolze$log2FoldChange.y<0]="Down"

cont_tb2 <- table(us=CommonGenesUsHolze$ourTest,them=CommonGenesUsHolze$theirTest) #Contingency table 2, for all groups
chisq.test(cont_tb2) #p< 2.2e-16
```

Heatmaps
```{r}
#Version without the names, for the genes with abs(log2FoldChange)>4 
select <- na.omit(resUsName$Row.names[resUsName$padj<0.05 & abs(resUsName$log2FoldChange)>4])
row.names(samples)=samples$run

pheatmap(assay(vsd)[select,], cluster_rows=TRUE, show_rownames=TRUE,show_colnames=FALSE,
         cluster_cols=FALSE, annotation_col=samples)

#version with the names 
select2 <- na.omit(resUsName$Row.names[resUsName$padj<0.05 & abs(resUsName$log2FoldChange)>4])
row.names(samples)=samples$run
resUsNameVsd <- merge(resUsName,assay(vsd),by.x="Row.names",by.y=0)
row.names(resUsNameVsd) <- resUsNameVsd$Row.names
h=pheatmap(resUsNameVsd[select2,10:15], cluster_rows=TRUE, show_rownames=TRUE,show_colnames=FALSE,
           cluster_cols=FALSE, annotation_col=samples, labels_row=resUsNameVsd[select,"external_gene_name"],fontsize_row=5)
h
ggsave(h,file="heatmapUpReg.png")

#Version for target selected genes 
cibles <- c("SAMD9L","SAMD9", "SAMHD1","RNASEL") #names of the genes of interest
pheatmap(resUsNameVsd[resUsNameVsd$external_gene_name%in%cibles,10:15], cluster_rows=TRUE, show_rownames=TRUE,show_colnames=FALSE,
       cluster_cols=FALSE, annotation_col=samples, labels_row=resUsNameVsd[resUsNameVsd$external_gene_name%in%cibles,"external_gene_name"])

```
